rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,299 @@
+  known interface that is declared in the service owner's specific Context type.
+
+## Registering new contexts
+
+```ts
+type ContextProvider<T extends keyof Context> = (...args: unknown[]) => Promise<Context[T]>;
+
+interface HandlerService {
+  registerContext<T extends keyof Context>(contextName: T, provider: ContextProvider<T>): void;
+}
+```
+
+- `args` in this example is specific to the handler type, for instance in a http
+  route handler, this would include the incoming request object. It would not
+  include the results from the other context providers in order to keep
+  providers from having dependencies on one another.
+- The `HandlerService` is defined as a literal interface in this document, but
+  in practice this interface is just a guide for the pattern of registering
+  context values. Certain services may have multiple different types of
+  handlers, so they may choose not to use the generic name `registerContext` in
+  favor of something more explicit.
+
+## End to end example
+
+```js
+http.router.registerRequestContext('elasticsearch', async request => {
+  const client = await core.elasticsearch.client$.toPromise();
+  return client.child({
+    headers: { authorization: request.headers.authorization },
+  });
+});
+
+http.router.route({
+  path: '/foo',
+  async routeHandler(context) {
+    context.elasticsearch.search(); // === callWithRequest(request, 'search')
+  },
+});
+```
+
+## Types
+
+While services that implement this pattern will not be able to define a static
+type, plugins should be able to reopen a type to extend it with whatever context
+it provides. This allows the `registerContext` function to be type-safe.
+For example, if the HTTP service defined a setup type like this:
+
+```ts
+// http_service.ts
+interface RequestContext {
+  core: {
+    elasticsearch: ScopedClusterClient;
+  };
+  [contextName: string]?: unknown;
+}
+
+interface HttpSetup {
+  // ...
+
+  registerRequestContext<T extends keyof RequestContext>(
+    contextName: T,
+    provider: (request: Request) => RequestContext[T] | Promise<RequestContext[T]>
+  ): void;
+
+  // ...
+}
+```
+
+A consuming plugin could extend the `RequestContext` to be type-safe like this:
+
+```ts
+// my_plugin/server/index.ts
+import { RequestContext } from '../../core/server';
+
+// The plugin *should* add a new property to the RequestContext interface from
+// core to represent whatever type its context provider returns. This will be
+// available to any module that imports this type and will ensure that the
+// registered context provider returns the expected type.
+declare module "../../core/server" {
+  interface RequestContext {
+    myPlugin?: {  // should be optional because this plugin may be disabled.
+      getFoo(): string;
+    }
+  }
+}
+
+class MyPlugin {
+  setup(core) {
+    // This will be type-safe!
+    core.http.registerRequestContext('myPlugin', (request) => ({
+      getFoo() { return 'foo!' }
+    }))
```

Comment:
I agree with @lukeelmers that some more non http examples might help, and I think I would take it a little farther. It throws me off a little bit that the `request` context is the same argument passed into the `registerRequestContext`  `ContextProvider` and the `routeHandler` function, and I'm not sure if that's by design or just a coincidence.  Like what does the Task manager tasks example from earlier, or the alerting one, look like once using this new API?   is the `http.router` one the only one that uses `args`?

I think my biggest open questions are:
 - how args fits in to both the provider and the handler function. Are they the same or totally unrelated?  I'm not sure I'm sold on this pattern vs passing context into functions provided on the plugin contract.  All examples I can think of seem a little more straightforward if you just use the plugin contracts. Like it's only one pattern I have to learn about.  
 - What `registerContext` functions look like.  I know this is an implementation detail but I think it'd help me understand where args comes from and some examples of args that aren't the `request` object.

This is a bad example but I'm trying to work out an alerting example on my own... can't really think of something that looks better on context vs plugin contract though.

```ts
// core setup
return {
  core: {
    alerting: {
      registerContext: (name: string, fn: (alertSettings: AlertSettings) => RequestContext[T] ) => {
        // store context providers somehow.  Does every service need to re-implement this function?
        // Is it executed now?, like
        alertingContextProviders[name] = fn(alertSettings);
        // or later prior to the handler being invoked?, like
        alertingContextProviders[name] = fn;
      }
   }
 } 
} 

// interpreter/plugin.ts
class InterpreterPlugin {
public setup(core: CoreSetup, plugins: PluginsSetup) {
  core.alerting.registerContext(('interpreter', (alertingSettings) => {
    return {
      interpret: (initialContext, expression) => this.interpret({ ...initialContext, ...alertingSettings }, expression);
    }
  });

  return { interpret }
};
}

// expression_alert/plugin.ts
class ExpressionAlertPlugin {
public setup(core: CoreSetup, plugins: PluginsSetup) {
  core.alerting.registerHandler({
  id: 'executeExpressionAlert',
  async handler(context, params, state) {
    const evaluatedExpression = context.interpreter.interpret(params.expression);
    if (evaluatedExpression > params.threshold) { fireAlert() }
  });
}
}
```

vs just doing:

```ts
// expression_alert/plugin.ts
class ExpressionAlertPlugin {
public setup(core: CoreSetup, plugins: PluginsSetup) {
  core.alerting.registerHandler({
  id: 'executeExpressionAlert',
  async handler(context, params, state, alertSettings) {
    const evaluatedExpression = plugins.interpreter.interpret({ ...params.initialContext, ...alertSettings }, params.expression);
    if (evaluatedExpression > params.threshold) { fireAlert() }
  });
}
}
```